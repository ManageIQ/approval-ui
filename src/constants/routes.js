const routes = {
  allrequests: {
    index: '/allrequests'
  },
  requests: {
    index: '/requests',
    addComment: '/requests/add-comment',
    approve: '/requests/approve',
    deny: '/requests/deny'
  },
  request: {
    index: '/request',
    addComment: '/request/add-comment',
    approve: '/request/approve',
    deny: '/request/deny'
  },
  workflows: {
    index: '/workflows',
    add: '/workflows/add-workflow',
    remove: '/workflows/remove',
    editInfo: '/workflows/edit-info',
    editGroups: '/workflows/edit-groups',
    editSequence: '/workflows/edit-sequence'
  }
};

export default routes;
