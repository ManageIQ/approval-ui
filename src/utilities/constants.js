import {
  CheckCircleIcon,
  InfoCircleIcon,
  ErrorCircleOIcon,
  ExclamationCircleIcon
} from '@patternfly/react-icons';
import React from 'react';
import requestsMessages from '../messages/requests.messages';

export const APPROVAL_API_BASE = `${process.env.BASE_PATH}/approval/v1.2`;
export const RBAC_API_BASE = `${process.env.BASE_PATH}/rbac/v1`;

export const decisionValues = {
  undecided: { displayName: requestsMessages.needsReview, color: 'blue', icon: <InfoCircleIcon /> },
  approved: { displayName: requestsMessages.approved, color: 'green', icon: <CheckCircleIcon /> },
  denied: { displayName: requestsMessages.denied, color: 'red', icon: <ExclamationCircleIcon /> },
  canceled: { displayName: requestsMessages.canceled, color: 'red', icon: <ErrorCircleOIcon /> },
  error: { displayName: requestsMessages.error, color: 'red', icon: <ExclamationCircleIcon /> }
};

// React intl does not support empty strings
export const untranslatedMessage = (defaultMessage = ' ') => ({ id: 'untranslated', defaultMessage });
export const APP_DISPLAY_NAME = {
  catalog: 'Automation Services Catalog',
  topology: 'Topological inventory'
};
