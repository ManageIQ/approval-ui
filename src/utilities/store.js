
import promiseMiddleware from 'redux-promise-middleware';
import ReducerRegistry, { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/files/ReducerRegistry';
import { notifications, notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import reduxLogger from 'redux-logger';
import thunk from 'redux-thunk';
import requestReducer, { requestsInitialState } from '../redux/reducers/request-reducer';
import workflowReducer, { workflowsInitialState } from '../redux/reducers/workflow-reducer';
import groupReducer, { groupsInitialState } from '../redux/reducers/group-reducer';
import rolesReducer, { rolesInitialState } from '../redux/reducers/roles-reducer';

const registry = new ReducerRegistry({}, [ thunk, promiseMiddleware(), notificationsMiddleware({
  errorTitleKey: [ 'message' ],
  errorDescriptionKey: [ 'errors', 'stack' ]
}), reduxLogger ]);

registry.register({
  requestReducer: applyReducerHash(requestReducer, requestsInitialState),
  workflowReducer: applyReducerHash(workflowReducer, workflowsInitialState),
  groupReducer: applyReducerHash(groupReducer, groupsInitialState),
  rolesReducer: applyReducerHash(rolesReducer, rolesInitialState),
  notifications
});

export default registry.getStore();
