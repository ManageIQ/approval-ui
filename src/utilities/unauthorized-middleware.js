import { approvalHistory } from '../router';

const unAuthorizedMiddleware = () => (dispatch) => (action) => {
  const nextAction = { ...action };
  if (action.type.match(/_REJECTED$/) && action?.payload?.redirect) {
    setTimeout(() => {
      approvalHistory.push(action.payload.redirect.pathname, {
        from: approvalHistory.location
      });
    });
    return;
  }

  return dispatch(nextAction);
};

export default unAuthorizedMiddleware;
