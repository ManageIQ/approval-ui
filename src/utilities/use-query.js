import { useLocation } from 'react-router-dom';

const useQuery = (requiredParams = []) => {
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  return [
    requiredParams.reduce(
      (acc, curr) => ({
        ...acc,
        [curr]: query.get(curr)
      }),
      {}
    ),
    search,
    query
  ];
};

export default useQuery;
