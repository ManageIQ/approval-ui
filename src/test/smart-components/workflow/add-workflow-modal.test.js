import React from 'react';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store' ;
import { Modal } from '@patternfly/react-core';
import { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { componentTypes } from '@data-driven-forms/react-form-renderer';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import { APPROVAL_API_BASE, RBAC_API_BASE } from '../../../utilities/constants';
import FormRenderer from '../../../smart-components/common/form-renderer';
import AddWorkflowModal from '../../../smart-components/workflow/add-workflow-modal';
import { validatorTypes } from '@data-driven-forms/react-form-renderer/dist/index';

describe('<AddWorkflowModal />', () => {
  let initialProps;
  let initialState;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;
  const ComponentWrapper = ({ store, children, workflowId }) => (
    <Provider store={ store }>
      <MemoryRouter initialEntries={ [ `workflows/edit/${workflowId}` ] }>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {

    };
    initialState = {
      groupReducer: {
        rbacGroups: [{
          label: 'foo',
          value: 'bar'
        }]
      },
      workflowReducer: {
        workflows: [{
          id: '123',
          name: 'Workflow'
        }]
      }
    };
    mockStore = configureStore(middlewares);
  });

  it('should render correctly', () => {
    const store = mockStore({});
    apiClientMock.get(`${APPROVAL_API_BASE}/workflows`, mockOnce({ body: { data: []}}));
    const wrapper = shallow(<ComponentWrapper store={ store }><AddWorkflowModal { ...initialProps } /></ComponentWrapper>).dive();

    setImmediate(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it('should create edit variant of workflow modal', done => {
    const store = mockStore(initialState);

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows`, mockOnce({
      body: {
        data: [{
          name: 'workflow',
          id: '123'
        }]
      }
    }));

    apiClientMock.get(`${RBAC_API_BASE}/groups/stage1/`, mockOnce({
      body: {
        name: 'Group1',
        id: 'stage1'
      }
    }));

    apiClientMock.get(`${RBAC_API_BASE}/groups/stage2/`, mockOnce({
      body: {
        name: 'Group2',
        id: 'stage2'
      }
    }));

    apiClientMock.get(`${APPROVAL_API_BASE}/workflows/123`, mockOnce({
      body: {
        name: 'workflow',
        id: '123',
        group_refs: [ 'stage1', 'stage2' ]
      }
    }));

    const expectedSchema = {
      fields: [{
        component: componentTypes.WIZARD,
        name: 'workflow_wizard',
        fields: [{
          title: 'Workflow information',
          name: 'wf_step_info',
          stepKey: 1,
          nextStep: 'wf_step_stages',
          fields: [{
            component: componentTypes.TEXTAREA_FIELD,
            name: 'name',
            type: 'text',
            isRequired: true,
            label: 'Name',
            validate: [{
              type: validatorTypes.REQUIRED
            }]
          }, {
            component: componentTypes.TEXTAREA_FIELD,
            name: 'description',
            type: 'text',
            label: 'Description'
          }]
        }, {
          stepKey: 'wf_step_stages',
          title: 'Set Approval Stages',
          name: 'wf_step_stages',
          nextStep: 'summary',
          component: componentTypes.FORM_GROUP,
          fields: [{
            name: 'wfGroups',
            component: componentTypes.FIELD_ARRAY,
            fields: [{
              component: componentTypes.SELECT,
              name: 'stage',
              isRequired: false,
              label: 'Stage',
              options: [{
                label: 'None',
                value: undefined
              }]
            }]
          }]
        }, {
          fields: [{
            name: 'summary',
            component: 'summary'
          }],
          stepKey: 'summary',
          name: 'summary'
        }]
      }]
    };

    const wrapper = mount(
      <ComponentWrapper store={ store } workflowId="123">
        <Route path="workflows/edit/:id?" render={ () => <AddWorkflowModal { ...initialProps } match={ { params: { id: '123' }} } /> }/>
      </ComponentWrapper>
    );

    setImmediate(() => {
      const modal = wrapper.find(Modal);
      const form = wrapper.find(FormRenderer);
      expect(modal.props().title).toEqual('Edit workflow');
      expect(form.props().schema).toEqual(expectedSchema);
      done();
    });
  });

  it('should create edit variant of workflow modal and call updateWorkflow on submit', () => {
    const store = mockStore(initialState);

    apiClientMock.patch(`${APPROVAL_API_BASE}/workflows/123`, ((req, res) => {
      expect(JSON.parse(req.body())).toEqual({ id: '123', name: 'Workflow' });
      return res.body(200);
    }));

    const wrapper = mount(
      <ComponentWrapper store={ store } workflowId="123">
        <Route
          path="workflows/edit/:id?"
          render={ () => <AddWorkflowModal { ...initialProps } match={ { params: { id: '123' }} }/> }
        />
      </ComponentWrapper>
    );

    const button = wrapper.find('button').last();
    button.simulate('click');
  });
});
