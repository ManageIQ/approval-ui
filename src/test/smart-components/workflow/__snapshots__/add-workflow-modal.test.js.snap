// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddWorkflowModal /> should create edit variant of workflow modal 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "workflows/edit/undefined",
    ]
  }
>
  <withRouter(Connect(AddWorkflowModal)) />
</MemoryRouter>
`;
